caliban_toolbox/reshape_data.py

```diff
@@ -387,3 +387,283 @@ def save_npzs_for_caliban(resized_xr, original_xr, log_data,  save_dir, blank_la
+
+                        # remove IDs and counts corresponding to overlap with ID 0 (background)
+                        keep_vals = np.nonzero(stitched_overlap_vals)
+                        stitched_overlap_vals = stitched_overlap_vals[keep_vals]
+                        stitched_overlap_counts = stitched_overlap_counts[keep_vals]
+
+                        # if there are overlaps, determine which is greatest in count, and replace with that ID
+                        if len(stitched_overlap_vals) > 0:
+                            max_overlap = stitched_overlap_vals[np.argmax(stitched_overlap_counts)]
+                            crop[crop == cell] = max_overlap
+
+                    # combine the crop with the current values in the stitched image
+                    combined_crop = np.where(stitched_crop > 0, stitched_crop, crop)
+
+                    # use this combined crop to update the values of stitched image
+                    stitched_labels[fov, stack, 0, 0, row_starts[row]:row_ends[row], col_starts[col]:col_ends[col], 0] = combined_crop
+
+                    crop_counter += 1
+
+    # relabel images to remove skipped cell_ids
+    return stitched_labels
+
+
+def reconstruct_image_stack(crop_dir, verbose=True):
+    """High level function to combine crops together into a single stitched image
+
+    Args:
+        crop_dir: directory where cropped files are stored
+        verbose: flag to control print statements
+
+    Returns:
+        None (saves stitched xarray to folder)
+    """
+
+    # sanitize inputs
+    if not os.path.isdir(crop_dir):
+        raise ValueError('crop_dir not a valid directory: {}'.format(crop_dir))
+
+    # unpack JSON data
+    with open(os.path.join(crop_dir, 'log_data.json')) as json_file:
+        log_data = json.load(json_file)
+
+    row_padding, col_padding = log_data['row_padding'], log_data['col_padding']
+    fov_names = log_data['fov_names']
+    # combine all npz crops into a single stack
+    crop_stack = load_npzs(crop_dir=crop_dir, log_data=log_data, verbose=verbose)
+
+    # stitch crops together into single contiguous image
+    stitched_images = stitch_crops(annotated_data=crop_stack, log_data=log_data)
+
+    # crop image down to original size
+    if row_padding > 0:
+        stitched_images = stitched_images[:, :, :, :, :-row_padding, :, :]
+    if col_padding > 0:
+        stitched_images = stitched_images[:, :, :, :, :, :-col_padding, :]
+
+    _, stack_len, _, _, row_len, col_len, _ = log_data['original_shape']
+    stitched_xr = xr.DataArray(data=stitched_images,
+                               coords=[fov_names, range(stack_len), range(1), range(1), range(row_len), range(col_len),
+                                       ['segmentation_label']], dims=['fovs', 'stacks', 'crops', 'slices',
+                                                                      'rows', 'cols', 'channels'])
```

Comment:
this is pretty complicated. maybe you can break out the logic a bit so someone could understand what's happening?